backend/src/main/java/de/bund/digitalservice/ris/norms/application/service/VerkuendungsImportService.java

```diff
@@ -47,4 +88,200 @@ public UUID storeNormendokumentationspaket(Query query) throws IOException {
+
+    while (!dokumenteToProcess.isEmpty()) {
+      String dokumentName = dokumenteToProcess.poll();
+      log.debug("Processing Dokument {}", dokumentName);
+      var dokument = parseAndValidateDokument(dokumentName, files.get(dokumentName));
+      dokumente.put(dokumentName, dokument);
+
+      if (!dokument.getManifestationEli().getSubtype().equals(dokumentName.split("\\.")[0])) {
+        throw new MismatchBetweenFilenameAndSubtypeException(
+          dokumentName,
+          dokument.getManifestationEli()
+        );
+      }
+
+      if (normManifestationEli == null) {
+        normManifestationEli = dokument.getManifestationEli().asNormEli();
+      }
+
+      if (!dokument.getManifestationEli().asNormEli().equals(normManifestationEli)) {
+        throw new InconsistentEliException(
+          dokumentName,
+          normManifestationEli,
+          dokument.getManifestationEli().asNormEli()
+        );
+      }
+
+      log.debug(
+        "Parsed & validated Dokument {} now looking for referenced dokuments in it",
+        dokument.getManifestationEli()
+      );
+
+      dokumenteToProcess.addAll(
+        dokument
+          .getReferencedDokumenteNames()
+          .stream()
+          // check if we already know about the referenced dokument
+          .filter(Predicate.not(dokumente::containsKey))
+          .filter(Predicate.not(dokumenteToProcess::contains))
+          .filter(referencedDokumentName -> {
+            if (isXmlFile(referencedDokumentName)) {
+              return true;
+            }
+            log.debug(
+              "Ignore referenced dokument {} as it is not an xml document.",
+              referencedDokumentName
+            );
+            return false;
+          })
+          .filter(referencedDokumentName -> {
+            if (files.containsKey(referencedDokumentName)) {
+              return true;
+            }
+            throw new MissingReferencedDokumentException(referencedDokumentName);
+          })
+          .peek(referencedDokumentName ->
+            log.debug("Found new referenced dokument {}", referencedDokumentName)
+          )
+          .collect(Collectors.toSet())
+      );
+    }
+
+    return new HashSet<>(dokumente.values());
+  }
+
+  private Dokument parseAndValidateDokument(String dokumentName, byte[] file) {
+    return switch (dokumentName.split("-")[0]) {
+      case "rechtsetzungsdokument" -> ldmlDeValidator.parseAndValidateRechtsetzungsdokument(
+        new String(file)
+      );
+      case "regelungstext" -> ldmlDeValidator.parseAndValidateRegelungstext(new String(file));
+      case "offenestruktur" -> ldmlDeValidator.parseAndValidateOffeneStruktur(new String(file));
+      case "bekanntmachungstext" -> ldmlDeValidator.parseAndValidateBekanntmachung(
+        new String(file)
+      );
+      default -> throw new InvalidDokumentTypeException(dokumentName);
+    };
+  }
+
+  private boolean isXmlFile(String dokumentName) {
```

Comment:
Also checking not 100% foolprof right? See comment about validating if ZIP is a ZIP

Reply:
this will need to be reworked after adding the rest of the import anyway as we probably "forget" the filenames before this step anyway, so I would leave it like this for now